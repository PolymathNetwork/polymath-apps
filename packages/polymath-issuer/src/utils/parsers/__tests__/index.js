import moment from 'moment';
import {
  validWhitelistCsv,
  invalidWhitelistCsv,
} from '../../../../testUtils/fixtures/csvFiles';
import { parseWhitelistCsv, validateWhitelistCsv } from '../index';

describe('parsers', () => {
  describe('parseWhitelistCsv', () => {
    test('parses the file correctly', () => {
      const result = parseWhitelistCsv(validWhitelistCsv);
      expect(result.data[0]).toEqual({
        address: '0x592E80AD45c08aba6C5bBd2d5C5A097BDF35Dee1',
        sellLockupDate: moment({ month: 0, day: 1, year: 2022 }).toDate(),
        buyLockupDate: moment({ month: 5, day: 4, year: 2024 }).toDate(),
        kycAmlExpiryDate: moment({ month: 0, day: 1, year: 2021 }).toDate(),
        canBuyFromSto: true,
        bypassesOwnershipRestriction: true,
        accredited: true,
        nonAccreditedLimit: null,
      });
      expect(result.invalidRows).toHaveLength(0);
    });

    test('returns a list of invalid rows', () => {
      const result = parseWhitelistCsv(invalidWhitelistCsv);
      expect(result.invalidRows).not.toHaveLength(0);
    });
  });

  describe('validateWhitelistCsv', () => {
    const sampleCsvRow = {
      address: '0x592E80AD45c08aba6C5bBd2d5C5A097BDF35Dee1',
      sellLockupDate: new Date('11/11/2018'),
      buyLockupDate: new Date('11/10/2018'),
      kycAmlExpiryDate: new Date('11/10/2022'),
      canBuyFromSto: true,
    };
    test('checks for invalid addresses', () => {
      const csv = [
        {
          ...sampleCsvRow,
          address: 'badAddress',
        },
      ];
      const invalidRows = validateWhitelistCsv(csv);
      expect(invalidRows).toHaveLength(1);
    });

    test('checks for duplicated addresses', () => {
      const csv = [
        {
          ...sampleCsvRow,
          address: '0x592E80AD45c08aba6C5bBd2d5C5A097BDF35Dee1',
        },
        {
          ...sampleCsvRow,
          address: '0x592E80AD45c08aba6C5bBd2d5C5A097BDF35Dee1',
        },
      ];

      const invalidRows = validateWhitelistCsv(csv);
      expect(invalidRows).toHaveLength(2);
    });

    test('checks for invalid valid dates', () => {
      const csv = [{ ...sampleCsvRow, buyLockupDate: 'notADate' }];
      const invalidRows = validateWhitelistCsv(csv);
      expect(invalidRows).toHaveLength(1);
    });

    test('checks for unexpected values', () => {
      const csv = [{ ...sampleCsvRow, isAccredited: 1234 }];
      const invalidRows = validateWhitelistCsv(csv);
      expect(invalidRows).toHaveLength(1);
    });
  });
});
