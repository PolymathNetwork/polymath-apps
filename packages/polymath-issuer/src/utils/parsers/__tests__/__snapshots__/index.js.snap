// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsers parseWhitelistCsv parses the file correctly 1`] = `
Object {
  "data": Array [
    Object {
      "accredited": true,
      "address": "0x592E80AD45c08aba6C5bBd2d5C5A097BDF35Dee1",
      "buyLockupDate": 2024-06-04T04:00:00.000Z,
      "bypassesOwnershipRestriction": true,
      "canBuyFromSto": true,
      "kycAmlExpiryDate": 2021-01-01T03:00:00.000Z,
      "nonAccreditedLimit": null,
      "sellLockupDate": 2022-01-01T03:00:00.000Z,
    },
    Object {
      "accredited": true,
      "address": "0xd9f346Bf88cA2cb7e11B0106018DE80A0169764D",
      "buyLockupDate": 2030-12-21T03:00:00.000Z,
      "bypassesOwnershipRestriction": null,
      "canBuyFromSto": true,
      "kycAmlExpiryDate": 2024-10-10T03:00:00.000Z,
      "nonAccreditedLimit": null,
      "sellLockupDate": 2030-12-21T03:00:00.000Z,
    },
    Object {
      "accredited": null,
      "address": "0xA1833Cd1a3e72335DE0b6945b5d83247F234d6e8",
      "buyLockupDate": 4099-01-01T00:00:00.000Z,
      "bypassesOwnershipRestriction": true,
      "canBuyFromSto": null,
      "kycAmlExpiryDate": 2025-12-01T03:00:00.000Z,
      "nonAccreditedLimit": 1000,
      "sellLockupDate": 4099-01-01T00:00:00.000Z,
    },
    Object {
      "accredited": null,
      "address": "0x07889A89C6854bb4Ec445825E680255b17751192",
      "buyLockupDate": 4099-01-01T00:00:00.000Z,
      "bypassesOwnershipRestriction": true,
      "canBuyFromSto": true,
      "kycAmlExpiryDate": 2025-12-01T03:00:00.000Z,
      "nonAccreditedLimit": null,
      "sellLockupDate": 4099-01-01T00:00:00.000Z,
    },
  ],
  "invalidRows": null,
}
`;
